In [1]:
#-*- coding: utf-8 -*-

# Analiza tekstu

## Zadanie

Napiszemy program, który czyta utwór literacki z pliku tekstowego i następnie dokonamy analizy tego utworu.

# Krok 1: Wczytanie nazwy pliku

Do czytania danych ze standardowego wejścia służy funkcja `input()`:
```python
input()
input('Podaj nazwę pliku')
```

In [ ]:
input()

# Krok 2: Sprawdzenie, czy plik istnieje

Najpierw musimy zaimportować moduł

```python
import os
```
Dzięki temu modułowi możemy sprawdzić, czy plik istnieje:
```python
os.path.exists(<ścieżka_do_pliku>)
```


Metoda zwraca wartość logiczną. Możemy więc skorzystać z instrukcji warunkowej:
   
```python
if warunek_logiczny_1:
    # instrukcje wykonywane, gdy warunek_logiczny_1 jest prawdziwy
elif warunek_logiczny_2:
    # instrukcje wykonywane, gdy warunek_logiczny_1 nie jest prawdziwy,
    # ale warunek_loginczy_2 jest prawdziwy
else:
    # instrukcje wykonywane, gdy ani warunek_logiczny_1 ani
    # warunek_logiczny_2 nie zachodzą
```

## Podsumowanie

Możemy napisać kod:

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")
if os.path.exists(file_path):
    print("Plik " + file_path + " istnieje")
else:
    print("Plik " + file_path + " NIE istnieje")

# Krok 3: Wymuszenie podania właściwego pliku

Program, w którym nie będziemy mieli ścieżki do istniejącego pliku, jest bezużyteczny. Musimy wymusić na użytkowniku podanie właściwej ścieżki. Możemy to zrobić przy użyciu pętli:

```python
while warunek:
    # instrukcje wykonywane dopóki warunek jest prawdziwy
```

## Podsumowanie

Poprzedni kod możemy przepisać:

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

# Krok 4: Czytanie pliku tekstowego

Do otwarcia pliku do czytania używamy instrukcji:
```python
f = open(<ścieżka do pliku>)
```
A gdy chcemy pisać do pliku:
```python
f = open(<ścieżka do pliku>, 'w')
```

## Sposób 1:

Możemy przeczytać cały plik do jednej zmiennej:

In [8]:
f = open('reduta-ordona.txt')
content = f.read()
print(r)

## Sposób 2:

Możemy czytać plik linijka po linijce:

In [18]:
f = open('reduta-ordona.txt')
line = f.readline()
while line != '':
    print(line)
    line = f.readline()

## Sposób 3:

Możemy przeczytać plik do listy linijek:

In [19]:
f = open('reduta-ordona.txt')
lines = f.readlines()

## Listy w pythonie

### Tworzenie listy:

In [22]:
l = [1, 2, 3.0, 'pawel', True]
print(l)

[1, 2, 3.0, 'pawel', True]


### Dostęp do dowolnego elementu w liście używając indeksu:

In [24]:
print(l[0])
print(l[3])
print(l[-1])

1
pawel
True


### Wybór podlisty:

In [25]:
print(l[:3])
print(l[2:])
print(l[1:-1])

[1, 2, 3.0]
[3.0, 'pawel', True]
[2, 3.0, 'pawel']


### Znalezienie długości listy

In [30]:
len(l)

5

### Powiększanie listy

In [40]:
print(l)
l.append(4)
print(l)
print(l + [3,2,1])

[1, 2, 3.0, 'pawel', True, 4, 4]
[1, 2, 3.0, 'pawel', True, 4, 4, 4]
[1, 2, 3.0, 'pawel', True, 4, 4, 4, 3, 2, 1]


### Sprawdzenie, czy element jest na liście

In [41]:
print(1 in l)
print('bogdan' in l)

True
False


### Przeglądanie całej listy

In [26]:
for element in l:
    print(element)

1
2
3.0
pawel
True


## Uwaga

Wszystko powyższe stosuje się do napisów. Napis należy rozumieć, jako listę liter

In [28]:
napis = "Paweł Bogdan"
print(napis)
print(napis[0])
print(napis[-1])
print(napis[:5])

Paweł Bogdan
P
n
Paweł


## Pętla `for each`

```python
for element in lista:
    # w każdym przebiegu pętli
    # w zmiennej element będzie kolejny element znajdujący się na liście
```

# Krok 5: policzenie częstotliwości występowania poszczególnych liter

Możemy skorzystać z funkcji `ord`:

In [33]:
print(ord('a'))
print(ord('z'))
print(ord('A'))
print(ord('Z'))

97
122
65
90


Tworzymy wtedy tablicę o rozmiarze 122 i jesteśmy w stanie w odpowiedniej komórce przechowywać ilość wystąpień liter o zadanych kodach ASCII.

## Wady?

### Dużo nieużywanych komórek, które zajmują miejsce w pamięci

### A co z innymi literami?

In [35]:
print(ord('ą'))
print(ord('Ź'))

261
377


### Potrzebujemy tablicy, w której indeksami może być coś innego niż kolejne liczby naturalne

## Rozwiązanie: słownik

### Tworzenie pustego słownika

In [36]:
d = {}

### Tworzenie słownika

In [37]:
d = {'a': 1, 'v': 5, 6: 'pawel', 4.5: 2}

### Dostęp do wartości:

In [43]:
print(d['a'])
print(d[6])

1
pawel


### Edycja słownika

In [44]:
d['a'] = 4
d['bogdan'] = 'pawel'

### Przeglądanie słownika

In [53]:
print(d)
print(d.keys())
print(d.values())
for k in d:
    print(d[k])
print(d.items())
print('a' in d)

{'a': 4, 'v': 5, 6: 'pawel', 4.5: 2, 'bogdan': 'pawel'}
dict_keys(['a', 'v', 6, 4.5, 'bogdan'])
dict_values([4, 5, 'pawel', 2, 'pawel'])
4
5
pawel
2
pawel
dict_items([('a', 4), ('v', 5), (6, 'pawel'), (4.5, 2), ('bogdan', 'pawel')])
True


## Dygresja - krotki

### Lista:

In [50]:
l = [2,3,4]
print(l)

[2, 3, 4]


### Krotka

(po angielsku: _tuple_)

In [51]:
k = (2,3,4)
print(k)

(2, 3, 4)


### Przeglądanie listy krotek

In [52]:
for k, v in d.items():
    print(k,v)

a 4
v 5
6 pawel
4.5 2
bogdan pawel


## Podsumowanie

In [55]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if not (letter in freq):
        freq[letter] = 1
    else:
        freq[letter] = freq[letter] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

Podaj ścieżkę do pliku: reduta-ordona.txt


# Krok 6: Wyeliminowanie niepotrzebnych znaków

Do statystyki liter nie są nam potrzebne białe znaki, znaki przystankowe i cyfry.

Do sprawdzenia, czy napis w zmiennej `s` zawiera tylko litery możemy wykorzystać wywołanie: `s.isalpha()`.


## Podsumowanie

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if letter.isalpha():
        if not (letter in freq):
            freq[letter] = 1
        else:
            freq[letter] = freq[letter] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

## Albo

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if not letter.isalpha():
        continue
    if not (letter in freq):
        freq[letter] = 1
    else:
        freq[letter] = freq[letter] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

# Krok 7: Brak rozróżnienia, czy litera jest wielka czy mała

Do sprawdzenia, czy napis w zmiennej `s` składa się tylko z wielkich liter można użyć polecenia `s.isupper()`, a czy składa się tylko z małych liter polecenia `s.islower()`.

Możemy też znaleźć odpowiednik napisu w zmiennej `s` w którym wszystkie litery są zamienione na wielkie przez użycie `s.upper()` i podobnie z małymi literami: `s.lower()`.

## Podsumowanie

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if not letter.isalpha():
        continue
    l = letter.lower()
    if not (l in freq):
        freq[l] = 1
    else:
        freq[l] = freq[l] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

# Krok 8: Posortowanie wyników